rs/execution_environment/src/execution/response.rs

```diff
@@ -518,17 +518,15 @@ impl ResponseHelper {
                 .get()
                 .saturating_sub(instructions_left.get()),
         );
+        self.canister
+            .system_state
+            .unregister_callback(original.callback_id)
+            .unwrap();
         let (action, call_context) = self
             .canister
             .system_state
-            .call_context_manager_mut()
-            .unwrap()
-            .on_canister_result(
-                original.call_context_id,
-                Some(original.callback_id),
-                result,
-                instructions_used,
-            );
+            .on_canister_result(original.call_context_id, result, instructions_used)
```

Comment:
I think that `CallContextManager:: on_canister_result()` can stay as it was even if we force it to be called from `SystemState::on_canister_result()` because the CallContextManager is now unreachable.

I would prefer to have the same arguments on `SystemState::on_canister_result()` as currently have `CallContextManager:: on_canister_result()` without adding `unregister_callback()` function. 

That way if we add implicit deletion of call context we can remove `CallContextID` argument from  `SystemState::on_canister_result()`. 

I believe that changing this function which has been in use for a long time and splitting it into `on_canister_result` and `unregister_callback` can create potential problems that someone who is used to how it works can forget to unregister context in the moment it has `CallContextId`.

Reply:
Oh, I see what you mean now (splitting `on_canister_result(..., callback_id, ...)` into `unregister_callback(calback_id)` plus `on_canister_result(...)`).

I can definitely roll back the change I made, although I would have argued the exact opposite. The `callback_id` argument to `on_canister_result()` was totally undocumented. One would have had to look at the implementation to understand why they shouldn't just pass `None`.

OTOH I fully agree with your argument that if we turn this `SystemState` into a state machine and forgo the need for an explicit `callback_id` argument, then a single `on_canister_result()` call would make more sense.

Done.

Reply:
Thank you.